### PR TITLE
Do not warn about shadowed fields if they are not redefined in a child class

### LINKED ISSUE #9107: UserWarning: Field name "XXX" shadows an attribute in parent "XXX"
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

When I put a field with an assigned default value into a separated mix-in class (for reuse), and include it in another Pydantic module, it generates the following unexpected and annoying warning.

```
(venv) imacat@host ~/project % ./test.py
/home/imacat/project/venv/lib/python3.11/site-packages/pydantic/_internal/_fields.py:186: UserWarning: Field name "my_field" shadows an attribute in parent "MyMixin"; 
  warnings.warn(
OK
(venv) imacat@host ~/project % 
```

### Example Code

```Python
#! env python3
from pydantic import BaseModel
class MyMixin:
    my_field: bool = False
class MyModel(BaseModel, MyMixin):
    pass
print("OK")
```

### Python, Pydantic & OS Version

```Text
(venv) imacat@host ~/project % python -c "import pydantic.version; print(pydantic.version.version_info())"
             pydantic version: 2.6.4
        pydantic-core version: 2.16.3
          pydantic-core build: profile=release pgo=true
                 install path: /home/imacat/project/venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0]
                     platform: Linux-6.1.0-18-amd64-x86_64-with-glibc2.36
             related packages: pydantic-settings-2.2.1 typing_extensions-4.10.0 fastapi-0.110.0
                       commit: unknown
(venv) imacat@host ~/project %
```

### PR DESCRIPTION
**Pull request by chan-vince**
## Change Summary

Fix for issue #9107.

Adds another early exit condition when evaluating whether to log a warning message during detection for shadowed fields.

In the case where a field is defined in a parent class, but it has not been defined _at all_ in a child class, it is technically not a shadowed field, and so shouldn't be warned as such.

Note this is very different from the case where a child class does redefine a field but with a narrower type or even defined as the same type but with a different default. Conceptually this is probably ok, but checking for that is quite complex and this PR does not attempt to try. So this is about checking if a field is defined or not defined - if it is, regardless of type or default value, the warning message is still logged.

Example of a case where a warning is currently logged, but is now no longer logged:

```python
class MyMixin:
    my_field: bool = False

class MyModel(BaseModel, MyMixin):
    pass
``` 

Example of a case where a warning is still logged, and it is important to do so:

```python
class MyMixin:
    my_field: bool = False

class MyModel(BaseModel, MyMixin):
    my_field: bool = True
```

## Related issue number

fix #9107

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @adriangb

### CODE REVIEW

**Comment by chan-vince** [comment]
please review

**Comment by sydney-runkle** [comment]
@chan-vince,

Looks great. Could you add a test for each of those above cases, showcasing that a warning is raised for one and not the other?

File: tests/test_main.py [hunk 1]
```diff
@@ -3151,6 +3152,33 @@ class Three(One):
...
+    """https://github.com/pydantic/pydantic/issues/9107"""
+
+    # A simple class which defines a field
+    class Parent:
+        foo: bool = False
+
+    # When inheriting from the parent class, as long as the field is not defined at all, there should be no warning
+    # about shadowed fields.
+    with warnings.catch_warnings():
+        warnings.simplefilter('error')
```
Comment (sydney-runkle) [inline tests/test_main.py:3165]: Why is this necessary?

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Looks good overall, just one comment!

File: tests/test_main.py (hunk 1 above)
Comment (chan-vince) [reply tests/test_main.py]: Sorry, I misunderstood the `simplefilter()` docs - I have adjusted it now.

Essentially follows [this example](https://docs.python.org/3/library/warnings.html#testing-warnings) from the docs, to ensure that all warnings are captured, then run the code, then check the recorded warnings list is empty.

Comment (sydney-runkle) [reply tests/test_main.py]: Ahh I see, looks good.

**Review by sydney-runkle — APPROVED** [review]
Thanks for adding the test!
